### PR TITLE
feat(echo.v4): add ignored routes env var

### PR DESCRIPTION
**Pull request by Trighap52**
## Summary

This adds an Echo v4-specific env var for skipping spans on configured routes:

`DD_TRACE_ECHO_IGNORED_ROUTES=GET /ready,POST /health`

The value holds comma-separated entries in the form `METHOD /pattern`. The middleware matches each entry against the request method and the route pattern registered by Echo. It does not match the request URL, so parameterized routes should use the registered pattern, for example `GET /users/:id` rather than `GET /users/123`.

Malformed non-empty entries are ignored and logged so configuration typos are visible. When a route is ignored, the Echo middleware returns before creating its span. This also skips AppSec monitoring performed by that middleware for the ignored route, matching the existing `WithIgnoreRequest` behavior.

This keeps the behavior in `dd-trace-go`, following the direction discussed in DataDog/orchestrion#788, and uses runtime configuration instead of source directives so Orchestrion users can configure ignored routes without source changes or rebuilds.

## Testing

- `go run ./scripts/configinverter/main.go check`
- `go test ./contrib/labstack/echo.v4`
- `go run github.com/DataDog/orchestrion go test ./echo.v4` from `internal/orchestrion/_integration`

Coverage includes:

- parser edge cases for empty/whitespace values, trailing commas, malformed entries, duplicate entries, and lower-case methods
- `Wrap` and direct `Middleware` setup paths
- grouped routes and parameterized route patterns
- Orchestrion auto-instrumentation with `DD_TRACE_ECHO_IGNORED_ROUTES=GET /health`, covering an ignored route and a traced route

Docs: DataDog/documentation#39746

Refs DataDog/orchestrion#788
Refs DataDog/orchestrion#709

### CODE REVIEW

**Review by rarguelloF — COMMENTED** [review]
Thanks for this, the health-check-ignoring use case makes sense. Before this lands, I'd like to discuss whether an env var would fit better than the `//dd:ignore-echo-route` directive.

The directive still requires editing source and rebuilding through Orchestrion for every route, which cuts against the main appeal of auto-instrumentation. An env var like `DD_TRACE_ECHO_IGNORED_ROUTES=GET /health,GET /ready` would let Orchestrion users exclude routes with zero source changes. The method+path matching in `isIgnoredRoute` would work the same either way.

One real trade-off: an env var can't tell apart two `echo.Echo` instances in the same process that register the same method+path if you want different behavior on each. The pointer-based `IgnoreRoute()` here handles that for free. Probably a rare setup, but worth naming.

Curious what you think, happy to help sketch out the env var version if that direction makes sense.

**Comment by Trighap52** [comment]
That makes sense to me. I initially went with the directive because it preserves precise route identity, including the multiple-`echo.Echo` case, but I agree the common health-check use case is better served as runtime configuration.

The env var approach also avoids the Orchestrion join-point edge cases around chained registrations like `e.Group("/api").GET(...)`, since matching happens after Echo resolves the route.

I can rework this PR around `DD_TRACE_ECHO_IGNORED_ROUTES=GET /health,GET /ready` and drop the directive unless you think there is still value in supporting both surfaces.

**Comment by rarguelloF** [comment]
@Trighap52 

> I can rework this PR around DD_TRACE_ECHO_IGNORED_ROUTES=GET /health,GET /ready and drop the directive unless you think there is still value in supporting both surfaces.

I think we can just go with the env variable. Will be happy to review once the new version is ready!

**Review by rarguelloF — CHANGES_REQUESTED** [review]
Let's change it to env variable instead as discussed here: https://github.com/DataDog/dd-trace-go/pull/5139#pullrequestreview-4895509098

**Review by darccio — CHANGES_REQUESTED** [review]
Agreed with @rarguelloF. Requesting changes too.

File: contrib/labstack/echo.v4/route.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/labstack/echo/v4"
+)
+
+func parseIgnoredRoutes(value string) map[string]struct{} {
+	ignoredRoutes := make(map[string]struct{})
+	for _, route := range strings.Split(value, ",") {
+		fields := strings.Fields(route)
+		if len(fields) != 2 {
+			continue
```
Comment (darccio) [inline contrib/labstack/echo.v4/route.go:26]: A malformed entry drops silently here. A user who types `GET /a b` or `GET` gets no ignored routes and no reason why. Log the dropped entry so the user can find the typo. The logger is available through `instr` in this package.

```suggestion
		if len(fields) != 2 {
			if strings.TrimSpace(route) != "" {
				instr.Logger().Warn("contrib/labstack/echo.v4: %s: entry %q is malformed. Each entry must hold one method and one route pattern, for example \"GET /ready\". The middleware keeps tracing malformed entries.", envIgnoredRoutes, route)
			}
			continue
		}
```

File: contrib/labstack/echo.v4/option.go [hunk 2]
```diff
@@ -16,8 +16,12 @@ import (
 	"github.com/labstack/echo/v4"
 )
 
-// envServerErrorStatuses is the name of the env var used to specify error status codes on http server spans
-const envServerErrorStatuses = "DD_TRACE_HTTP_SERVER_ERROR_STATUSES"
+const (
+	// envServerErrorStatuses is the name of the env var used to specify error status codes on http server spans.
+	envServerErrorStatuses = "DD_TRACE_HTTP_SERVER_ERROR_STATUSES"
+	// envIgnoredRoutes is the name of the env var used to specify route resources whose spans should be skipped.
```
Comment (darccio) [inline contrib/labstack/echo.v4/option.go]: The env var comment does not state the entry format. The tests show the format, but the tests are not documentation. State the format, one example, and the match rule at the const.

```suggestion
	// envIgnoredRoutes is the name of the env var used to skip spans for routes.
	// The value holds comma-separated entries in the form "METHOD /pattern".
	// Example: "GET /ready,POST /health".
	// The middleware matches each entry against the request method and the route pattern that Echo registered. It does not match the request URL.
	envIgnoredRoutes = "DD_TRACE_ECHO_IGNORED_ROUTES"
```

File: contrib/labstack/echo.v4/route.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+package echo
+
+import (
+	"strings"
+
+	"github.com/labstack/echo/v4"
+)
+
+func parseIgnoredRoutes(value string) map[string]struct{} {
```
Comment (darccio) [inline contrib/labstack/echo.v4/route.go:14]: `parseIgnoredRoutes` has no unit tests. The three middleware tests cover the happy path only. Please add unit tests for these cases:

- an empty value returns nil
- a value with spaces only returns nil
- a trailing comma does not create an entry
- an entry with one field or three fields is dropped
- a duplicate entry does not change the result
- a lower-case method, for example `get /ready`, matches an upper-case request method

File: contrib/labstack/echo.v4/route.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		return nil
+	}
+	return ignoredRoutes
+}
+
+func isIgnoredRoute(cfg *config, c echo.Context) bool {
+	if cfg == nil || len(cfg.ignoredRoutes) == 0 || c == nil || c.Request() == nil || c.Path() == "" {
+		return false
+	}
+	_, found := cfg.ignoredRoutes[routeKey(c.Request().Method, c.Path())]
```
Comment (darccio) [inline contrib/labstack/echo.v4/route.go]: `parseIgnoredRoutes` uppercases the method from the env value, but this lookup compares `c.Request().Method` as-is. A request with a lower-case method does not match an entry. Go servers receive upper-case methods, so this case is rare. Uppercase the request method so the match holds for any case.

```suggestion
	_, found := cfg.ignoredRoutes[routeKey(strings.ToUpper(c.Request().Method), c.Path())]
```

File: contrib/labstack/echo.v4/echotrace.go [hunk 5]
```diff
@@ -62,7 +62,7 @@ func Middleware(opts ...Option) echo.MiddlewareFunc {
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
 			// If we have an ignoreRequestFunc, use it to see if we proceed with tracing
-			if cfg.ignoreRequestFunc != nil && cfg.ignoreRequestFunc(c) {
+			if isIgnoredRoute(cfg, c) || (cfg.ignoreRequestFunc != nil && cfg.ignoreRequestFunc(c)) {
```
Comment (darccio) [inline contrib/labstack/echo.v4/echotrace.go:65]: The ignored path returns before the handler creates the span and before it wraps `next` with `withAppSec`. So an ignored route loses WAF coverage as well as the span. An attacker can probe an ignored route without AppSec events.

The `ignoreRequestFunc` path has the same effect, so the behavior is consistent with the existing API. But the primary use case for this env var is a health check, so users will apply it to routes that stay exposed to the network.

Two options:

1. Accept the behavior and state the consequence in the env var docs: "The middleware also skips AppSec monitoring for ignored routes."
2. Keep AppSec for ignored routes, and skip only the span.

Which option do you prefer?

Comment (darccio) [inline contrib/labstack/echo.v4/echotrace.go]: The comment names `ignoreRequestFunc` only. The condition also checks the route set from the env var. A reader who trusts the comment misses the env var path. Update the comment to cover both cases.

```suggestion
			// Skip tracing when an ignored route matches, or when the ignoreRequestFunc does.
			if isIgnoredRoute(cfg, c) || (cfg.ignoreRequestFunc != nil && cfg.ignoreRequestFunc(c)) {
```

File: contrib/labstack/echo.v4/echotrace_test.go [hunk 6]
```diff
@@ -547,6 +547,113 @@ func TestIgnoreRequestFunc(t *testing.T) {
...
+	})
+
+	readyReq := httptest.NewRequest(http.MethodGet, "/api/ready", nil)
+	readyRec := httptest.NewRecorder()
+	router.ServeHTTP(readyRec, readyReq)
+
+	assert.Equal(http.StatusNoContent, readyRec.Code)
+	assert.True(readyCalled)
+	assert.Len(mt.FinishedSpans(), 0)
+}
```
Comment (darccio) [inline contrib/labstack/echo.v4/echotrace_test.go:738]: The tests do not cover a param pattern. The middleware matches the registered pattern, not the request URL, so a route like `/users/:id` matches only when the env entry holds the pattern with the placeholder. This is the most likely user mistake, because the request URL contains `123` where the pattern holds `:id`. Add one test for the param case.

```suggestion
}

func TestIgnoredRoutesEnvWithParamRoute(t *testing.T) {
	t.Setenv(envIgnoredRoutes, "GET /users/:id")
	assert := assert.New(t)
	mt := mocktracer.Start()
	defer mt.Stop()

	var userCalled bool

	router := Wrap(echo.New())
	router.GET("/users/:id", func(c echo.Context) error {
		userCalled = true
		return c.NoContent(http.StatusNoContent)
	})

	userReq := httptest.NewRequest(http.MethodGet, "/users/123", nil)
	userRec := httptest.NewRecorder()
	router.ServeHTTP(userRec, userReq)

	assert.Equal(http.StatusNoContent, userRec.Code)
	assert.True(userCalled)
	assert.Len(mt.FinishedSpans(), 0)
}
```

**Review by darccio — CHANGES_REQUESTED** [review]
The Orchestrion integration suite for Echo (`internal/orchestrion/_integration/echo.v4`) does not exercise the env var. Auto-instrumentation is the primary audience for `DD_TRACE_ECHO_IGNORED_ROUTES`, because the reviewers chose the env var to let those users skip routes without source changes.

Please add one scenario to that suite:

1. Set `DD_TRACE_ECHO_IGNORED_ROUTES` before the instrumented server starts.
2. Register a route such as `GET /health`.
3. Assert that a request to `/health` produces no span, and a request to a traced route produces one span.

The middleware code path is the same as the manual path, so this scenario adds confidence, not correctness.

The PR description describes the previous design. It names the `//dd:ignore-echo-route` directive, the `IgnoreRoute` helper, and the Orchestrion join points. None of these exist in the head. The description also lists three Orchestrion test commands that no longer apply.

Please rewrite the description for the env var design. State these points:

- The env var `DD_TRACE_ECHO_IGNORED_ROUTES` and one example value: `GET /ready,POST /health`.
- The entry format: one method and one route pattern, comma-separated.
- The match rule: the middleware matches the request method and the route pattern that Echo registered. It does not match the request URL.
- The tests that cover the feature.

This matters because the team treats the PR description as the record of the change.

The env var has no user-facing documentation in this repo. The supported-configurations entry and the const comment are the only places that name it. The docs team documents tracer env vars outside this repo. Does a ticket exist for `DD_TRACE_ECHO_IGNORED_ROUTES` in the external docs? If yes, please link it here. If no, please open one, because a user who learns the feature from docs also needs the entry format and the match rule.

Conclusion: the final design is an improvement over the first version in three ways. The feature works for `Wrap` users and for plain `Middleware` users, because the check lives in the one shared code path. The matching runs at request time, so chained group routes such as `e.Group("/api").GET("/ready")` work without a directive whitelist. The global route registry is gone, so no package-level state exists to leak across Echo instances. Please keep this shape in future contribs that gain the same feature.

File: contrib/labstack/echo.v4/echotrace.go (hunk 5 above)
Comment (Trighap52) [reply contrib/labstack/echo.v4/echotrace.go:65]: I chose option 1 for this PR: ignored routes keep the same behavior as `WithIgnoreRequest`, so the Echo middleware returns before creating the span and before AppSec wrapping.

I documented that consequence in the PR description and included it in the docs issue: DataDog/documentation#39746. Keeping AppSec while skipping only the Echo span seems like a larger behavior change that should be handled separately.

**Review by darccio — APPROVED** [review]
LGTM. Thanks for the effort, @Trighap52!

**Review by rarguelloF — APPROVED** [review]
